### DISCUSSION TITLE
Subscriptions - change document in subscription without triggering subscription again

### DISCUSSION THREAD

**Discussion by ben-jacobs** [ravendb/ravendb#13485] [Q&A] [answered]
Hi,

I know this case is _sort of_ covered in one of the examples where the subscription checks for a null in a property the persists a value in the subscription ([Subscription that works with a session](https://ravendb.net/docs/article-page/5.3/csharp/client-api/data-subscriptions/consumption/examples#subscription-that-works-with-a-session)). 

This is a pretty rudimentary case however - in real-world scenarios, subscriptions might be used to put a message in a queue to run, for example, a set of business rules. These business rules might, for examples, patch the document to enrich the document with data sourced from elsewhere. In this case, you wouldn't want to retrigger the rules from running again when persisting the changes in the patch (circular updates etc).

There's potential solutions using multi-hop approaches or the business rules could just patch/update/create a separate document linked to the main document (this is actually what I do but it's not really nice). The problem with this latter case is that I'm playing with a VERY large dataset (hundreds of millions of documents) so the metadata alone for the extra documents is a killer. The example also doesn't work as is as I'd want to rerun the rules on every document change OTHER than the changes made within the subscription itself.

Any thoughts on how this could be achieved other than the way shown in the examples? I considered using revisions to check if the only change in the last update was to the dynamic area and skip queue the message in that case - but this still results in the document being put through the subscription twice. However I think about trying to handle the situation I end up with either a lot of complexity, or a lot of extra IO/storage etc for what would be really simple if I could somehow veto the subscription for running on a single save.

Cheers,

Ben

**Comment by ayende** [comment, marked as answer]
The problem is not a simple one. You need two separate scenarios to deal with. You have one time processing and a multiple processing scenarios.

In most cases, subscriptions are used on the domain documents where there is inherently a single reason to make the change.
If you have a scenario where you want to respond to the same document from a subscription potentially multiple times, my recommendation would be to split the logic.

Basically, you have the document that matters to you, and when you want to invoke a subscription, don't rely on that document for this.
Create a separate collection dedicated just for the subscription notification, and delete the extra document as part of handling the document batch.

That gives you exactly what you want, with great simplicity.  You can also `include` the source document on the subscription, so you'll not have to pay additional network roundtrips.

We'll likely make that a first class feature in future releases, see:




See: https://issues.hibernatingrhinos.com/issue/RavenDB-17865

**Comment by ben-jacobs** [comment]
I like where you're heading with your thinking in the feature outline. 

If the message/document is deleted as part of the batch, what would the interaction be if there were multiple subscriptions listening for the notification (or would this be a singleton)? 

Would this cause issues if the second listener were significantly delayed? i.e. would the document be deleted only after acknowledged by all subscriptions etc?

**Reply by ayende** [reply]
If you have multiple subscriptions, you create a separate message doc for each of those.
That way, they can operate completely independently.
